emulator/periph/src/flash_ctrl.rs

```diff
@@ -0,0 +1,1096 @@
+/*++
+
+Licensed under the Apache-2.0 license.
+
+File Name:
+
+    flash_ctrl.rs
+
+Abstract:
+
+    File contains dummy flash controller peripheral emulation.
+
+--*/
+
+use core::convert::TryInto;
+use emulator_bus::{Clock, ReadOnlyRegister, ReadWriteRegister, Timer};
+use emulator_cpu::Irq;
+use emulator_derive::Bus;
+use emulator_registers_generated::flash::FlashPeripheral;
+use registers_generated::flash_ctrl::bits::{
+    Control, CtrlRegwen, InterruptEnable, InterruptState, OpStatus,
+};
+use std::fs::File;
+use std::io::{Read, Seek, Write};
+use std::path::PathBuf;
+use tock_registers::interfaces::{ReadWriteable, Readable, Writeable};
+
+#[derive(Debug, PartialEq)]
+#[allow(clippy::enum_variant_names)]
+pub enum FlashOperation {
+    ReadPage = 1,
+    WritePage = 2,
+    ErasePage = 3,
+}
+
+impl TryInto<FlashOperation> for u32 {
+    type Error = ();
+
+    fn try_into(self) -> Result<FlashOperation, Self::Error> {
+        match self {
+            1 => Ok(FlashOperation::ReadPage),
+            2 => Ok(FlashOperation::WritePage),
+            3 => Ok(FlashOperation::ErasePage),
+            _ => Err(()),
+        }
+    }
+}
+
+pub enum FlashCtrlIntType {
+    Error = 1,
+    Event = 2,
+}
+
+pub enum FlashOpError {
+    ReadError = 0,
+    WriteError = 1,
+    EraseError = 2,
+    InvalidOp = 3,
+}
+
+// Define a struct to represent the flash controller peripheral emulation.
+// Includes IRQs, timers, and peripheral registers, backend storage and a buffer page_size for the flash controller.
+#[derive(Bus)]
```

Comment:
If you implement `FlashPeripheral`, you shouldn't need to use `derive(Bus)` -- the `FlashPeripheral` wrapper in the `AutoRootBus` will take care of the `Bus` methods for you.

That means you can probably remove the `ReadWriteRegister<u32>`, etc., below if you want and just use `u32` or `Cell<u32>` (for example)

Reply:
Hi @swenson, thanks for catching this issue.  
Do we prefer normal variable `u32` to `ReadWriteRegister<u32>`  in this context?
It seems that `ReadWriteRegister<u32>` etc. still works by importing `use emulator_bus::{ReadOnlyRegister, ReadWriteRegister};` event without `derive(bus)`.  Please advise if I have to refactor it to `u32`.

Reply:
I would say use whichever is easier for you.